google/cloud/forseti/services/inventory/base/cai_gcp_client.py

```diff
@@ -109,65 +109,65 @@ def session(self):
-            project_number, dataset_id)
-
-    def iter_bigquery_datasets(self, project_number):
-        """Iterate Datasets from Cloud Asset data.
-
-        Args:
-            project_number (str): number of the project to query.
-
-        Yields:
-            dict: Generator of datasets.
-        """
-        resources = self.dao.iter_cai_assets(
-            ContentTypes.resource,
-            'google.bigquery.Dataset',
-            '//cloudresourcemanager.googleapis.com/projects/{}'.format(
-                project_number),
-            self.session)
-        for dataset in resources:
-            yield dataset
+    # def fetch_bigquery_iam_policy(self, project_number, dataset_id):
+    #     """Gets IAM policy of a bigquery dataset from Cloud Asset data.
+    #
+    #     Args:
+    #         project_number (str): number of the project to query.
+    #         dataset_id (str): id of the dataset to query.
+    #
+    #     Returns:
+    #         dict: Dataset IAM Policy.
+    #     """
+    #     resource = self.dao.fetch_cai_asset(
+    #         ContentTypes.iam_policy,
+    #         'google.bigquery.Dataset',
+    #         '//bigquery.googleapis.com/projects/{}/datasets/{}'.format(
+    #             project_number, dataset_id),
+    #         self.session)
+    #     if resource:
+    #         return resource
+    #     return {}
+    #
+    # def fetch_bigquery_dataset_policy(self, project_number, dataset_id):
+    #     """Dataset policy Iterator for a dataset from Cloud Asset data.
+    #
+    #     Args:
+    #         project_number (str): number of the project to query.
+    #         dataset_id (str): id of the dataset to query.
+    #
+    #     Returns:
+    #         dict: Dataset Policy.
+    #     """
+    #     resource = self.dao.fetch_cai_asset(
+    #         ContentTypes.iam_policy,
+    #         'google.bigquery.Dataset',
+    #         '//bigquery.googleapis.com/projects/{}/datasets/{}'.format(
+    #             project_number, dataset_id),
+    #         self.session)
+    #     if resource:
+    #         return iam_helpers.convert_iam_to_bigquery_policy(resource)
+    #     # Fall back to live API if the data isn't in the CAI cache.
+    #     return super(CaiApiClientImpl, self).fetch_bigquery_dataset_policy(
+    #         project_number, dataset_id)
+    #
+    # def iter_bigquery_datasets(self, project_number):
+    #     """Iterate Datasets from Cloud Asset data.
+    #
+    #     Args:
+    #         project_number (str): number of the project to query.
+    #
+    #     Yields:
+    #         dict: Generator of datasets.
+    #     """
+    #     resources = self.dao.iter_cai_assets(
+    #         ContentTypes.resource,
+    #         'google.bigquery.Dataset',
+    #         '//cloudresourcemanager.googleapis.com/projects/{}'.format(
+    #             project_number),
+    #         self.session)
+    #     for dataset in resources:
```

Comment:
Here you can have a check to see if the dataset has location, return that, otherwise use the data from the API:

resources = ...
if resources and not all('location' in ds for ds in resources):
    LOGGER.info('Datasets missing location key in CAI, falling back to live API.')
    resources = super(CaiApiClientImpl, self).iter_bigquery_datasets(project_number)

for dataset in resources:
  yield dataset

Reply:
Updated, also casted resources to be a list instead of a generator, otherwise the check "not all('location' in ds for ds in resources)" will empty out the generator and the yield in the for loop afterward won't have any results.